### PR TITLE
Reduce instruction count for CH2 metadata ops.

### PR DESCRIPTION
**Pull request by arjunroy**
1) Statically pre-compute static slice indices to save some ALU ops when linking
batched metadata.
2) Change some asserts to debug_asserts since they can provably not be triggered
with the current implementation.

No API changes are made here.

Fullstack unary pingpong sees slight improvements:

BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4096/0                                                      [polls/iter:3.00009                        ]            25.2µs ± 3%             24.4µs ± 2%  -3.33%          (p=0.038 n=6+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/32768/0                                                     [polls/iter:3.0001                         ]            35.2µs ± 2%             34.3µs ± 2%  -2.41%          (p=0.017 n=6+5)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/262144/0                                                    [polls/iter:3.00011                        ]             154µs ± 1%              152µs ± 1%  -1.37%          (p=0.003 n=8+5)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/64                                                    [polls/iter:3.0001                         ]            23.3µs ± 2%             22.7µs ± 2%  -2.61%          (p=0.017 n=6+5)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/0                                                   [polls/iter:0                              ]            6.47µs ± 5%             6.32µs ± 3%  -2.37%        (p=0.006 n=20+17)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/1                                                   [polls/iter:0                              ]            6.61µs ± 4%             6.43µs ± 1%  -2.59%        (p=0.010 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/1/1                                                   [polls/iter:0                              ]            6.76µs ± 4%             6.56µs ± 1%  -2.96%        (p=0.011 n=20+18)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/64/0                                                  [polls/iter:0                              ]            6.63µs ± 4%             6.51µs ± 1%  -1.81%        (p=0.002 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/64                                                  [polls/iter:0                              ]            6.72µs ± 5%             6.48µs ± 1%  -3.67%        (p=0.000 n=20+18)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/64/64                                                 [polls/iter:0                              ]            6.79µs ± 4%             6.64µs ± 1%  -2.25%        (p=0.001 n=20+20)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/512/0                                                 [polls/iter:0                              ]            6.87µs ± 3%             6.67µs ± 1%  -2.85%        (p=0.000 n=20+18)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/512                                                 [polls/iter:0                              ]            6.90µs ± 4%             6.64µs ± 1%  -3.89%        (p=0.000 n=20+20)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/512/512                                               [polls/iter:0                              ]            7.08µs ± 4%             6.93µs ± 1%  -2.06%        (p=0.013 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/4096/0                                                [polls/iter:0                              ]            7.85µs ± 3%             7.67µs ± 1%  -2.24%        (p=0.000 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/4096                                                [polls/iter:0                              ]            7.89µs ± 4%             7.63µs ± 1%  -3.23%        (p=0.000 n=20+18)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/4096/4096                                             [polls/iter:0                              ]            9.05µs ± 3%             8.87µs ± 1%  -1.94%        (p=0.000 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/32768/0                                               [polls/iter:0                              ]            15.9µs ± 1%             15.7µs ± 1%  -1.00%        (p=0.001 n=19+20)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/32768                                               [polls/iter:0                              ]            16.0µs ± 2%             15.7µs ± 1%  -1.42%        (p=0.000 n=19+20)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/32768/32768                                           [polls/iter:0                              ]            25.1µs ± 1%             25.0µs ± 1%  -0.63%        (p=0.000 n=19+20)
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/0                                              [polls/iter:0                              ]            79.1µs ± 0%             78.8µs ± 1%  -0.39%        (p=0.001 n=20+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/0                                                [polls/iter:0                              ]            6.47µs ± 5%             6.39µs ± 6%  -1.14%        (p=0.038 n=20+20)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/1/0                                                [polls/iter:0                              ]            6.58µs ± 5%             6.45µs ± 1%  -1.97%        (p=0.000 n=20+20)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/1                                                [polls/iter:0                              ]            6.60µs ± 5%             6.44µs ± 1%  -2.41%        (p=0.001 n=20+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/1/1                                                [polls/iter:0                              ]            6.74µs ± 4%             6.55µs ± 1%  -2.89%        (p=0.000 n=20+20)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/8/0                                                [polls/iter:0                              ]            6.53µs ± 3%             6.45µs ± 1%  -1.26%        (p=0.013 n=20+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/8                                                [polls/iter:0                              ]            6.51µs ± 4%             6.41µs ± 1%  -1.55%        (p=0.041 n=20+18)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/8/8                                                [polls/iter:0                              ]            6.64µs ± 4%             6.55µs ± 1%  -1.26%        (p=0.048 n=20+18)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/64/0                                               [polls/iter:0                              ]            6.60µs ± 3%             6.53µs ± 2%  -1.08%        (p=0.001 n=18+18)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/64                                               [polls/iter:0                              ]            6.64µs ± 3%             6.51µs ± 1%  -2.02%        (p=0.000 n=20+20)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/512                                              [polls/iter:0                              ]            6.80µs ± 5%             6.67µs ± 1%  -1.95%        (p=0.000 n=20+20)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/4096/0                                             [polls/iter:0                              ]            7.74µs ± 3%             7.67µs ± 1%  -0.91%        (p=0.005 n=18+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/32768/0                                            [polls/iter:0                              ]            15.9µs ± 2%             15.7µs ± 1%  -0.76%        (p=0.000 n=19+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/32768                                            [polls/iter:0                              ]            15.9µs ± 2%             15.7µs ± 1%  -1.13%        (p=0.000 n=19+18)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/262144/0                                           [polls/iter:0                              ]            79.1µs ± 1%             78.7µs ± 0%  -0.55%        (p=0.000 n=20+19)
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/262144                                           [polls/iter:0                              ]            79.0µs ± 1%             78.7µs ± 1%  -0.31%        (p=0.011 n=20+19)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<10>, 1>, NoOpMutator>/0/0               [polls/iter:0                              ]            7.34µs ± 4%             7.21µs ± 4%  -1.70%        (p=0.013 n=20+20)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<100>, 1>, NoOpMutator>/0/0              [polls/iter:0                              ]            7.36µs ± 5%             7.26µs ± 2%  -1.33%        (p=0.014 n=20+20)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<10>, 2>, NoOpMutator>/0/0               [polls/iter:0                              ]            7.97µs ± 4%             7.82µs ± 2%  -1.96%        (p=0.005 n=20+20)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<31>, 2>, NoOpMutator>/0/0               [polls/iter:0                              ]            8.11µs ± 3%             7.97µs ± 1%  -1.73%        (p=0.000 n=20+17)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<100>, 2>, NoOpMutator>/0/0              [polls/iter:0                              ]            8.22µs ± 4%             8.02µs ± 1%  -2.44%        (p=0.000 n=20+19)
BM_UnaryPingPong<InProcess, NoOpMutator, Server_AddInitialMetadata<RandomBinaryMetadata<31>, 1>>/0/0        [polls/iter:0                              ]            7.25µs ± 4%             7.16µs ± 1%  -1.13%        (p=0.032 n=19+19)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomAsciiMetadata<31>, 1>, NoOpMutator>/0/0                [polls/iter:0                              ]            7.34µs ± 4%             7.26µs ± 1%  -1.09%        (p=0.014 n=20+20)
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomAsciiMetadata<100>, 1>, NoOpMutator>/0/0               [polls/iter:0                              ]            7.48µs ± 3%             7.38µs ± 1%  -1.30%        (p=0.006 n=20+20)
BM_UnaryPingPong<InProcess, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<10>, 1>>/0/0         [polls/iter:0                              ]            7.15µs ± 2%             7.11µs ± 1%  -0.44%        (p=0.041 n=16+19)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                         [polls/iter:3.00008                        ]            22.3µs ± 3%             21.0µs ± 1%  -5.56%          (p=0.036 n=5+3)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/512/0                                                       [polls/iter:3.0001                         ]            23.4µs ± 1%             22.9µs ± 1%  -2.14%          (p=0.029 n=4+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/32768/32768                                                 [polls/iter:3.0001                         ]            48.4µs ± 0%             47.8µs ± 1%  -1.42%          (p=0.016 n=4+5)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/0                                                      [polls/iter:3.00009                        ]            21.6µs ± 2%             20.9µs ± 1%  -3.12%          (p=0.036 n=5+3)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/0                                                      [polls/iter:3.00006                        ]            20.8µs ± 1%             20.3µs ± 2%  -2.42%          (p=0.024 n=3+6)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/64                                                     [polls/iter:3.00009                        ]            22.0µs ± 1%             21.5µs ± 0%  -1.92%          (p=0.036 n=3+5)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/64                                                    [polls/iter:3.00008                        ]            23.4µs ± 2%             22.4µs ± 1%  -4.10%          (p=0.036 n=3+5)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/32768/0                                                  [polls/iter:3.00012                        ]            33.8µs ± 0%             33.1µs ± 1%  -1.94%          (p=0.036 n=3+5)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                         [polls/iter:3.00011                        ]            22.5µs ± 2%             21.3µs ± 1%  -5.26%          (p=0.029 n=4+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4096/4096                                                   [polls/iter:3.00008                        ]            28.3µs ± 4%             27.0µs ± 2%  -4.78%          (p=0.038 n=4+6)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/1                                                      [polls/iter:3.00008                        ]            22.3µs ± 4%             21.3µs ± 1%  -4.59%          (p=0.048 n=6+3)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/1/0                                                         [polls/iter:3.00011                        ]            22.5µs ± 2%             21.8µs ± 1%  -3.20%          (p=0.016 n=4+5)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4096/4096                                                   [polls/iter:3.0001                         ]            27.6µs ± 2%             26.9µs ± 1%  -2.62%          (p=0.029 n=4+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/64/0                                                        [polls/iter:3.0001                         ]            23.4µs ± 3%             22.5µs ± 1%  -3.86%          (p=0.010 n=6+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/4096                                                      [polls/iter:3.00009                        ]            25.4µs ± 3%             24.1µs ± 1%  -4.75%          (p=0.016 n=4+5)
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/32768/0                                                  [polls/iter:3.00005                        ]            34.6µs ± 7%             33.2µs ± 1%  -3.89%          (p=0.016 n=5+4)
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/32768/32768                                                 [polls/iter:3.00007                        ]            48.6µs ± 1%             47.7µs ± 1%  -1.84%          (p=0.024 n=3+6)

### CODE REVIEW

**Comment by arjunroy** [comment]
Rebased to fix merge conflict.

File: src/core/lib/transport/static_metadata.h [hunk 1]
```diff
@@ -521,11 +524,16 @@ typedef union {
   } named;
 } grpc_metadata_batch_callouts;
 
-#define GRPC_BATCH_INDEX_OF(slice)                                        \
-  (GRPC_IS_STATIC_METADATA_STRING((slice))                                \
-       ? static_cast<grpc_metadata_batch_callouts_index>(                 \
-             GPR_CLAMP(GRPC_STATIC_METADATA_INDEX((slice)), 0,            \
-                       static_cast<intptr_t>(GRPC_BATCH_CALLOUTS_COUNT))) \
+#define GRPC_BATCH_INDEX_OF(slice)                                            \
+  (GRPC_IS_STATIC_METADATA_STRING((slice))                                    \
```
Comment (hcaseyal) [inline src/core/lib/transport/static_metadata.h]: Can we rewrite this so it's more readable? E.g.,:

GRPC_IS_STATIC_METADATA_STRING(slice) && reinterpret_cast<grpc_core::StaticSliceRefcount*>((slice).refcount)->index <= static_cast<uint32_t>(GRPC_BATCH_CALLOUTS_COUNT) ? static_cast<grpc_metadata_batch_callouts_index>(reinterpret_cast<grpc_core::StaticSliceRefcount*>(     (slice).refcount)->index)) : GRPC_BATCH_CALLOUTS_COUNT

Comment (arjunroy) [reply src/core/lib/transport/static_metadata.h]: Done.

File: tools/codegen/core/gen_static_metadata.py [hunk 2]
```diff
@@ -594,8 +594,25 @@ def f(i, p=p):
 print >> H, '  } named;'
 print >> H, '} grpc_metadata_batch_callouts;'
 print >> H
-print >> H, '#define GRPC_BATCH_INDEX_OF(slice) \\'
-print >> H, '  (GRPC_IS_STATIC_METADATA_STRING((slice)) ? (reinterpret_cast<grpc_core::StaticSliceRefcount*>((slice).refcount)->index > static_cast<uint32_t>(GRPC_BATCH_CALLOUTS_COUNT) ?  GRPC_BATCH_CALLOUTS_COUNT : static_cast<grpc_metadata_batch_callouts_index>(reinterpret_cast<grpc_core::StaticSliceRefcount*>((slice).refcount)->index)) : GRPC_BATCH_CALLOUTS_COUNT)'
+
+batch_idx_of_hdr = '#define GRPC_BATCH_INDEX_OF(slice) \\'
+static_slice = 'GRPC_IS_STATIC_METADATA_STRING((slice))'
+slice2sref = '(slice).refcount'
```
Comment (hcaseyal) [inline tools/codegen/core/gen_static_metadata.py]: Let's rename these so a future reader can understand them:

sref -> slice_ref
2 -> to
scast -> static_cast
inval -> invalid

E.g., slice2sref -> slice_to_slice_ref

Comment (arjunroy) [reply tools/codegen/core/gen_static_metadata.py]: Will do; those were my hacks to get things to fit in 80 char lines :P

Comment (arjunroy) [reply tools/codegen/core/gen_static_metadata.py]: Done.

Comment (hcaseyal) [reply tools/codegen/core/gen_static_metadata.py]: Ah I see...do you think that my requested changes combined with the reformatting will make things less readable? As it stands, I had to do some mental work and looking around to figure out what the variable names meant, which is not great. But I'm not sure if the 80 char requirement combined with the rename will make things worse.

Forgot to add: rcast , like scast, probably doesn't need to be a variable.

Comment (hcaseyal) [reply tools/codegen/core/gen_static_metadata.py]: Never mind, just saw your changes and they look good.

File: tools/codegen/core/gen_static_metadata.py [hunk 3]
```diff
@@ -595,8 +594,25 @@ def f(i, p=p):
...
 print >> H, '} grpc_metadata_batch_callouts;'
 print >> H
-print >> H, '#define GRPC_BATCH_INDEX_OF(slice) \\'
-print >> H, '  (GRPC_IS_STATIC_METADATA_STRING((slice)) ? static_cast<grpc_metadata_batch_callouts_index>(GPR_CLAMP(GRPC_STATIC_METADATA_INDEX((slice)), 0, static_cast<intptr_t>(GRPC_BATCH_CALLOUTS_COUNT))) : GRPC_BATCH_CALLOUTS_COUNT)'
+
+batch_idx_of_hdr = '#define GRPC_BATCH_INDEX_OF(slice) \\'
+static_slice = 'GRPC_IS_STATIC_METADATA_STRING((slice))'
+slice2sref = '(slice).refcount'
+static_sref_type = 'grpc_core::StaticSliceRefcount*'
+rcast = 'reinterpret_cast'
```
Comment (soheilhy) [inline tools/codegen/core/gen_static_metadata.py]: please spell these out in places they happen. you can go above the 80-column limit here.

Comment (soheilhy) [reply tools/codegen/core/gen_static_metadata.py]: I meant basically removing `scast` and `rcast`.

Comment (arjunroy) [reply tools/codegen/core/gen_static_metadata.py]: Already fixed the names in response to @hcaseyal .

Regarding why I'm using these short variable names at all - for the really cumbersome stuff like the casts, the final join() statement that assembles things is a lot easier to parse without the awful C++ syntax gumming it up.

File: src/core/ext/transport/chttp2/transport/incoming_metadata.cc [hunk 4]
```diff
@@ -38,7 +38,8 @@ grpc_error* grpc_chttp2_incoming_metadata_buffer_add(
     storage = static_cast<grpc_linked_mdelem*>(
         buffer->arena->Alloc(sizeof(grpc_linked_mdelem)));
   }
-  return grpc_metadata_batch_add_tail(&buffer->batch, storage, elem);
+  storage->md = elem;
```
Comment (yashykt) [inline src/core/ext/transport/chttp2/transport/incoming_metadata.cc]: any special reason for this? Are we rolling out all functions like these?

Comment (arjunroy) [reply src/core/ext/transport/chttp2/transport/incoming_metadata.cc]: Putting aside my confusion that both methods exist - I like keeping small operations like setting a qword to be on the near side of a compilation unit barrier, giving the compiler the freedom to move things around if needed during optimization. Not saying it makes a huge difference in this case, though.

**Pull request merged**